### ISSUE TITLE
Performance issue on getCanvasBBox() during graph.render()

### ISSUE THREAD

**Issue by sklinov** [antvis/G6#2557]
- [x] I have searched the [issues](https://github.com/antvis/g6/issues) of this repository and believe that this is not a duplicate.

### Reproduction link
Sorry, but can't really reproduce that: (client's data) and custom layout

### Steps to reproduce
 But on a relatively small graph with some 1700 nodes (`modelRect`), 30 combos (`rect`), and 6300 edges (`cubic-horizontal`) `graph.render()` takes around 60 seconds to finish.
After that all operations with graph do not experience that issue.
Profiling in Firefox shows that it spends most of the time in `getCanvasBBox()` 
![image](https://user-images.githubusercontent.com/38185977/104710045-6ee62780-5741-11eb-996f-9c6e400a4561.png)

I've already set `localRefresh, fitView, fitCenter` to `false` to avoid calling those functions as much as possible (because it would crash).

Any ideas on what can be causing that slow-down and how to prevent that?

| Environment | Info |
|---|---|
| g6 | 4.1.1 |
| System | Windows 10 |
| Browser | Chrome 87.0.4280.141 |

---
Sorry, but can't really reproduce that: (client's data) and custom layout

**Comment by Yanyan-Wang** [comment]
I think most of the `getCanvasBBox()` is called by the combos since the size and the position of each combo depends on the sub-items' bounding box, including sub nodes and sub combos. And the combos construct a hierarchy tree. Maybe just try remove the combos to see whether it is the reason. If so, we might need to figure out how to optimize the combo mechanism in G6.

And `localRefresh` might not really relates to the problem. But for the first time rendering, turns `localRefresh` to be false might help. And turning it to be true after `graph.render` will be helpful for local interaction and refresh for better local rendering performance.

**Comment by sklinov** [comment]
@Yanyan-Wang Thanks for that comment.
My investigation now showed that it was `imageMinimap` plugin that was slowing the render down.
When I commented out the 
```
 graph.on("aftergraphrefresh", () => {
    graph.toFullDataURL(img => {
    ...
```
piece of code - performance changed dramatically (1.5 seconds for `graph.render()` instead of 38 seconds in my sample).
My guess would be that this event is fired somewhere in between render process?

**Comment by sklinov** [comment]
@Yanyan-Wang By the way, turning `localRefresh` on after the initial render still makes me run into `maximum call stack exceeded` error on `getCanvasBBox()` on large graphs, so I kept it `off` position

**Comment by Yanyan-Wang** [comment]
> @Yanyan-Wang By the way, turning `localRefresh` on after the initial render still makes me run into `maximum call stack exceeded` error on `getCanvasBBox()` on large graphs, so I kept it `off` position

That is strange, it will be helpful if you can provide a reproduction.

**Comment by sklinov** [comment]
> > @Yanyan-Wang Thanks for that comment.
> > My investigation now showed that it was `imageMinimap` plugin that was slowing the render down.
> > When I commented out the
> > ```
> >  graph.on("aftergraphrefresh", () => {
> >     graph.toFullDataURL(img => {
> >     ...
> > ```
> > 
> > 
> > piece of code - performance changed dramatically (1.5 seconds for `graph.render()` instead of 38 seconds in my sample).
> > My guess would be that this event is fired somewhere in between render process?
> 
> Ahh, `toFullDataURL` copies the whole canvas onto another one for getting the full picture of the graph. It will be better to generate the image for image minimap in pre-processing progress or in on the server but not on the front-end. If there is no server in your project, what about using minimap with keyShape type instead?

You mean moving the whole `G6` on the server side wrapping it with something like `node-canvas`?
Our graph is not static, it has some interactions that change the way it looks, so we have to regenerate minimap every time graph is updated.

**Comment by sklinov** [comment]
> > @Yanyan-Wang By the way, turning `localRefresh` on after the initial render still makes me run into `maximum call stack exceeded` error on `getCanvasBBox()` on large graphs, so I kept it `off` position
> 
> That is strange, it will be helpful if you can provide a reproduction.

Yeah, I'll try to reproduce that as close as possible.

**Comment by Yanyan-Wang** [comment]
> > > @Yanyan-Wang Thanks for that comment.
> > > My investigation now showed that it was `imageMinimap` plugin that was slowing the render down.
> > > When I commented out the
> > > ```
> > >  graph.on("aftergraphrefresh", () => {
> > >     graph.toFullDataURL(img => {
> > >     ...
> > > ```
> > > 
> > > 
> > > piece of code - performance changed dramatically (1.5 seconds for `graph.render()` instead of 38 seconds in my sample).
> > > My guess would be that this event is fired somewhere in between render process?
> > 
> > 
> > Ahh, `toFullDataURL` copies the whole canvas onto another one for getting the full picture of the graph. It will be better to generate the image for image minimap in pre-processing progress or in on the server but not on the front-end. If there is no server in your project, what about using minimap with keyShape type instead?
> 
> You mean moving the whole `G6` on the server side wrapping it with something like `node-canvas`?
> Our graph is not static, it has some interactions that change the way it looks, so we have to regenerate minimap every time graph is updated.

Nope, what I mean is the Image Minimap might be more suitable for a graph with static data, and the image of the minimap can be generated only once when the static data is outputted. If the data is not static, you have to generate the image frequently, which will be expensive.

So, maybe just try the Minimap with delegate type.

**Comment by Yanyan-Wang** [comment]
Hi, dear user. Since there was no response for a long time, we cannot figure out the answer of the problem you described. If there is new progress, feel free to open a new issue.
